bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,626 @@
+		tsr := trf.TimestampRangeFilter
+
+		tsf := bigtable.TimestampRangeFilter(time.UnixMicro(tsr.StartTimestampMicros), time.UnixMicro(tsr.EndTimestampMicros))
+		f = &tsf
+
+	case *btpb.RowFilter_ValueRegexFilter:
+		vrf := fpb.(*btpb.RowFilter_ValueRegexFilter)
+		re := vrf.ValueRegexFilter
+		vr := bigtable.ValueFilter(string(re))
+		f = &vr
+
+	case *btpb.RowFilter_ValueRangeFilter:
+		vrf := fpb.(*btpb.RowFilter_ValueRangeFilter)
+
+		var start, end []byte
+		switch sv := vrf.ValueRangeFilter.StartValue; sv.(type) {
+		case *btpb.ValueRange_StartValueOpen:
+			start = sv.(*btpb.ValueRange_StartValueOpen).StartValueOpen
+		case *btpb.ValueRange_StartValueClosed:
+			start = sv.(*btpb.ValueRange_StartValueClosed).StartValueClosed
+		}
+
+		switch ev := vrf.ValueRangeFilter.EndValue; ev.(type) {
+		case *btpb.ValueRange_EndValueOpen:
+			end = ev.(*btpb.ValueRange_EndValueOpen).EndValueOpen
+		case *btpb.ValueRange_EndValueClosed:
+			end = ev.(*btpb.ValueRange_EndValueClosed).EndValueClosed
+		}
+
+		vr := bigtable.ValueRangeFilter(start, end)
+		f = &vr
+
+	case *btpb.RowFilter_CellsPerRowOffsetFilter:
+		cof := fpb.(*btpb.RowFilter_CellsPerRowOffsetFilter)
+		off := cof.CellsPerRowOffsetFilter
+		co := bigtable.CellsPerRowOffsetFilter(int(off))
+		f = &co
+
+	case *btpb.RowFilter_CellsPerRowLimitFilter:
+		clf := fpb.(*btpb.RowFilter_CellsPerRowLimitFilter)
+		lim := clf.CellsPerRowLimitFilter
+		cl := bigtable.CellsPerRowLimitFilter(int(lim))
+		f = &cl
+
+	case *btpb.RowFilter_CellsPerColumnLimitFilter:
+		ccf := fpb.(*btpb.RowFilter_CellsPerColumnLimitFilter)
+		lim := ccf.CellsPerColumnLimitFilter
+		cc := bigtable.LatestNFilter(int(lim))
+		f = &cc
+
+	case *btpb.RowFilter_StripValueTransformer:
+		sv := bigtable.StripValueFilter()
+		f = &sv
+
+	case *btpb.RowFilter_ApplyLabelTransformer:
+		alf := fpb.(*btpb.RowFilter_ApplyLabelTransformer)
+		l := alf.ApplyLabelTransformer
+		al := bigtable.LabelFilter(l)
+		f = &al
+	default:
+		return nil
+	}
+
+	return f
+}
+
+// testClient contains a bigtable.Client object, cancel functions for the calls
+// made using the client, an appProfileID (optionally), and a
+// perOperationTimeout (optionally).
+type testClient struct {
+	c                   *bigtable.Client   // c stores the Bigtable client under test
+	appProfileID        string             // appProfileID is currently unused
+	perOperationTimeout *duration.Duration // perOperationTimeout sets a custom timeout for methods calls on this client
+}
+
+// credentialsBundle implements credentials.Bundle interface
+// [See documentation for usage](https://pkg.go.dev/google.golang.org/grpc/credentials#Bundle).
+type credentialsBundle struct {
+	channel credentials.TransportCredentials
+	call    credentials.PerRPCCredentials
+}
+
+// TransportCredentials gets the channel credentials as TransportCredentials
+func (c credentialsBundle) TransportCredentials() credentials.TransportCredentials {
+	return c.channel
```

Comment:
Should this be a pointer receiver? As written all the methods here will make a copy of channel and call, which might be OK, but just want to confirm.

Reply:
No, the interface requires a value receiver.